zcash_client_backend/src/proto.rs

```diff
@@ -183,3 +204,223 @@ impl service::TreeState {
+                    }
+                };
+
+                let transaction_request =
+                    TransactionRequest::from_uri(params, &self.transaction_request)?;
+                let transparent_inputs = self
+                    .transparent_inputs
+                    .iter()
+                    .map(|t_in| {
+                        let txid = t_in.parse_txid().map_err(ProposalError::TxIdInvalid)?;
+                        let outpoint = OutPoint::new(txid.into(), t_in.index);
+
+                        wallet_db
+                            .get_unspent_transparent_output(&outpoint)
+                            .map_err(ProposalError::InputRetrieval)?
+                            .ok_or_else(|| {
+                                ProposalError::InputNotFound(
+                                    txid,
+                                    PoolType::Transparent,
+                                    t_in.index,
+                                )
+                            })
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                let sapling_inputs = self.sapling_inputs.as_ref().and_then(|s_in| {
+                    s_in.inputs
+                        .iter()
+                        .map(|s_in| {
+                            let txid = s_in.parse_txid().map_err(ProposalError::TxIdInvalid)?;
+
+                            wallet_db
+                                .get_spendable_sapling_note(&txid, s_in.index)
+                                .map_err(ProposalError::InputRetrieval)
+                                .and_then(|opt| {
+                                    opt.ok_or_else(|| {
+                                        ProposalError::InputNotFound(
+                                            txid,
+                                            PoolType::Shielded(ShieldedProtocol::Sapling),
+                                            s_in.index,
+                                        )
+                                    })
+                                })
+                        })
+                        .collect::<Result<Vec<_>, _>>()
+                        .map(|notes| {
+                            NonEmpty::from_vec(notes).map(|notes| {
+                                SaplingInputs::from_parts(s_in.anchor_height.into(), notes)
+                            })
+                        })
+                        .transpose()
+                });
+
+                let balance = self.balance.as_ref().ok_or(ProposalError::BalanceInvalid)?;
+                let balance = TransactionBalance::new(
+                    balance
+                        .proposed_change
+                        .iter()
+                        .filter_map(|cv| {
+                            cv.value
+                                .as_ref()
+                                .map(|cv| -> Result<ChangeValue, ProposalError<_>> {
```

Comment:
When can `cv.value` be `None`? That would imply a protobuf encoding of a non-empty sequence of default enums? I feel like this should also be rejected.

Non-blocking because this is purely a representational non-canonicity, and thus is closer to the non-canonicity in the rest of protobuf (vs my earlier comment which points to a more structural issue).

Reply:
All object-valued types are `Option`al in protobuf encodings, so we have to handle that case. If there's no proposed change value for an entry, we can also treat that as an encoding error.

Reply:
Yep, that was my non-blocking conclusion above.